crates/rattler_installs_packages/src/wheel_builder.rs

```diff
@@ -0,0 +1,370 @@
+//! Turn an sdist into a wheel by creating a virtualenv and building the sdist in it
+use std::{
+    cell::RefCell,
+    collections::{HashMap, HashSet},
+    path::PathBuf,
+    process::{Command, Output},
+    rc::Rc,
+    str::FromStr,
+};
+
+use pep508_rs::{MarkerEnvironment, Requirement};
+
+#[derive(Debug)]
+/// A fake tempdir for debugging (doesn't delete stuff after it's done)
+struct FakeTempDir {
+    path: PathBuf,
+}
+
+impl FakeTempDir {
+    /// Create a new fake tempdir
+    pub fn new() -> Self {
+        let dir = tempfile::tempdir().unwrap();
+        Self {
+            path: dir.into_path(),
+        }
+    }
+
+    /// Get the path of the tempdir
+    pub fn path(&self) -> &PathBuf {
+        &self.path
+    }
+}
+
+use crate::{
+    core_metadata::{WheelCoreMetaDataError, WheelCoreMetadata},
+    resolve,
+    sdist::SDist,
+    tags::WheelTags,
+    venv::{PythonLocation, VEnv},
+    wheel::UnpackError,
+    Artifact, PackageDb, ResolveOptions, SDistName, SDistResolution, UnpackWheelOptions, Wheel,
+};
+
+#[derive(Debug)]
+pub struct CacheValue {
+    folder: FakeTempDir,
+    package_dir: PathBuf,
+    #[allow(dead_code)]
+    build_system: pyproject_toml::BuildSystem,
+    entry_point: String,
+    requirements: Vec<Requirement>,
+    venv: VEnv,
+}
+
+type BuildCache = RefCell<HashMap<SDistName, Rc<CacheValue>>>;
+
+// include static build_frontend.py string
+const BUILD_FRONTEND_PY: &str = include_str!("./wheel_builder_frontend.py");
+/// A builder for wheels
+pub struct WheelBuilder<'db, 'i> {
+    /// A cache for virtualenvs that might be reused later in the process
+    venv_cache: BuildCache,
+
+    /// The package database to use
+    package_db: &'db PackageDb,
+
+    /// The env markers to use when resolving
+    env_markers: &'i MarkerEnvironment,
+
+    /// The configured wheel tags to use when resolving
+    wheel_tags: Option<&'i WheelTags>,
+
+    /// The resolve options. Note that we change the sdist resolution to normal if it's set to
+    /// only sdists, because otherwise we run into a chicken & egg problem where a sdist is required
+    /// to build a sdist. E.g. `hatchling` requires `hatchling` as build system.
```

Comment:
Is this example in the commenct correct? It seems hatchling has wheel files?

Reply:
Oh sorry I get it now, it's the case for *only* sdists :)

Reply:
yep!